fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+        checkpointer: Optional[TorchCheckpointer] = None,
+        lam: float = 1.0,
+        mkmmd_loss_weight: float = 10.0,
+        beta_update_interval: int = 20,
+        feature_l2_norm: Optional[float] = 0.0,
+    ) -> None:
+        """
+        This client implements the Ditto algorithm from Ditto: Fair and Robust Federated Learning Through
+        Personalization. The idea is that we want to train personalized versions of the global model for each client.
+        So we simultaneously train a global model that is aggregated on the server-side and use those weights to also
+        constrain the training of a local model. The constraint for this local model is identical to the FedProx loss.
+
+        Args:
+            data_path (Path): path to the data to be used to load the data for client-side training
+            metrics (Sequence[Metric]): Metrics to be computed based on the labels and predictions of the client model
+            device (torch.device): Device indicator for where to send the model, batches, labels etc. Often 'cpu' or
+                'cuda'
+            loss_meter_type (LossMeterType, optional): Type of meter used to track and compute the losses over
+                each batch. Defaults to LossMeterType.AVERAGE.
+            checkpointer (Optional[TorchCheckpointer], optional): Checkpointer to be used for client-side
+                checkpointing. Defaults to None.
+            metrics_reporter (Optional[MetricsReporter], optional): A metrics reporter instance to record the metrics
+                during the execution. Defaults to an instance of MetricsReporter with default init parameters.
+            lam (float, optional): weight applied to the Ditto drift loss. Defaults to 1.0.
+            mkmmd_loss_weight (float, optional): weight applied to the MK-MMD loss. Defaults to 10.0.
+            beta_update_interval (int, optional): interval at which to update the betas for the MK-MMD loss.
+                Defaults to 20.
+            feature_l2_norm (Optional[float], optional): weight applied to the L2 norm of the features.
+            Defaults to 0.0.
+        """
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.lam = lam
+        self.mkmmd_loss_weight = mkmmd_loss_weight
+        self.feature_l2_norm = feature_l2_norm
+        self.beta_update_interval = beta_update_interval
+        self.mkmmd_loss = MkMmdLoss(device=self.device, minimize_type_two_error=True).to(self.device)
+
+        self.global_model: nn.Module
+        self.init_global_model: nn.Module
+
+    def get_optimizer(self, config: Config) -> Dict[str, Optimizer]:
+        """
+        Returns a dictionary with global and local optimizers with string keys 'global' and 'local' respectively.
+        """
+        raise NotImplementedError
```

Comment:
I know I didn't do this when I implemented Ditto originally, but I think it would be helpful for us to add a string to the reporting of this error along the lines of
```
NotImplementedError("User Clients must define a function that returns a Dict[str, Optimizer] with keys \"global\" and \"local\" defining separate optimizers for the global and local models of Ditto.")
```